Feature/recon documentation (#395)
databrickslabs/remorph: README_RECON.md

# Remorph Reconciliation

Remorph Reconciliation is a library that helps to reconcile the data, post data migration or during the data migration   
between the source system and databricks system.It supports multiple types of reconciliation depending upon the user   

> requirements and data in the source system.It makes the migration process simpler by removing the complex part of

Reviewer: move this under docs as recon.md and reference this inside main Readme.md

Author: done

validating the data from one system to another and enabling the team with more insights on the data migrated.

## Installation

## Flow Chart for supported Report Type and Reconciliation Logics

```mermaid
flowchart TD
    REPORT_TYPE --> DATA
    REPORT_TYPE --> SCHEMA
    REPORT_TYPE --> ROW
    REPORT_TYPE --> ALL
```

```mermaid
flowchart TD
    SCHEMA --> SCHEMA_VALIDATION
```

```mermaid
flowchart TD
    ROW --> MISSING_IN_SRC
    ROW --> MISSING_IN_TGT
```

```mermaid
flowchart TD
   DATA --> MISMATCH_ROWS
   DATA --> MISSING_IN_SRC
   DATA --> MISSING_IN_TGT
```

```mermaid
flowchart TD
    ALL --> MISMATCH_ROWS
    ALL --> MISSING_IN_SRC
    ALL --> MISSING_IN_TGT
    ALL --> SCHEMA_VALIDATION
```

## Supported Source System and Report Types

| Source     | Schema | Row | Data | All |
|------------|--------|-----|------|-----|
| Oracle     | Yes    | Yes | Yes  | Yes |
| Snowflake  | Yes    | Yes | Yes  | Yes |
| Databricks | Yes    | Yes | Yes  | Yes |

## Dependency CLI commands

> TBD

Reviewer: CLI will trigger Jobs, @bishwajit-db or @vijaypavann-db who is working on the cli piece?

## Environment Setup

> TBD

Reviewer: this will not be separate for recon but rather consistent for remorph.

Author: removed

## TABLE Config Elements:

> | config_name         | data_type    | description                                                                                                                                                                                                                      | required/optional | example_value                                                                                                                  |

Reviewer: This table is Great,mention how to create them as dataclassess as well.

| thresholds          | list         | list of threshold conditions, that can be applied on tgt columns to match the minor exceptions in data.It supports percentile,absolute, and date fields.For more info [thresholds](#thresholds)                                  | optional          | "thresholds": [{"column_name": "sal","lower_bound": "-5%","upper_bound": "5%","type": "integer"}]                              |
| filters             | string       | filter expr that can be used to filter the data on src and tgt based on respective expressions                                                                                                                                   | optional          | "filters" : {"source":"lower(dept_name)>'it'" ,"target" : "lower(department_name)>'it'"}                                       |

### jdbc_reader_options

> | field_name        | data_type | description                                                                                                                                                                                                                                                                                                                                                                                                                                                            | required/optional | example_value |

Reviewer: Same as above

| lower_bound       | string    | integer or date or timestamp without time zone value as string), that should be set appropriately (usually the minimum value in case of non-skew data) so the data read from the source should be approximately equally distributed                                                                                                                                                                                                                                    | required          | "100000"      |
| fetch_size        | string    | This parameter influences the number of rows fetched per round-trip between Spark and the JDBC database, optimizing data retrieval performance. Adjusting this option significantly impacts the efficiency of data extraction, controlling the volume of data retrieved in each fetch operation. More details on configuring fetchSize can be found [here](https://docs.databricks.com/en/connect/external-systems/jdbc.html#control-number-of-rows-fetched-per-query) | optional          | "10000"       |

### column_mapping

| field_name  | data_type | description        | required/optional | example_value   |

> |-------------|-----------|--------------------|-------------------|-----------------|

Reviewer: Same as above

| source_name | string    | source column name | required          | "dept_id"       |
| target_name | string    | target column name | required          | "department_id" |

### transformations

| field_name  | data_type | description                                                | required/optional | example_value                    |

> |-------------|-----------|------------------------------------------------------------|-------------------|----------------------------------|

Reviewer: same as above

| column_name | string    | the column name on which the transformation to be applied  | required          | "s_address"                      |
| source      | string    | the transformation sql expr to be applied on source column | required          | "trim(s_address)" or "s_address" |
| target      | string    | the transformation sql expr to be applied on source column | required          | "trim(s_address)" or "s_address" |

### thresholds

| field_name  | data_type | description                                                                                                                                                                  | required/optional                  | example_value      |

> |-------------|-----------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------------------------|--------------------|

Reviewer: Same as above

| column_name | string    | the column that should be considered for threshold reconciliation                                                                                                            | required                           | "product_discount" |
| lower_bound | string    | the lower bound of the difference between the source value and the target value                                                                                              | optional(default 0)                | -5%                | 
| upper_bound | string    | the upper bound of the difference between the source value and the target value                                                                                              | optional(default 0)                | 5%                 |            
| type        | string    | user must specify the type of the column. integer is selected for columns of type integer or double, while timestamp is chosen for columns of either date or timestamp types | required("integer" or "timestamp") | "integer"          |

### Key Considerations on Table Config:
1. The column names are always converted to lower case and considered for reconciliation.

> 2. Currently it doesn't support case sensitive column names

Reviewer: ```suggestion
2. Currently it doesn't support case insensitivity and doesn't have collation support
```

Author: done

3. Table Transformation internally consider the default value as the column value. It doesn't apply any default transformations
if not provided.
```eg:Transformation(column_name="address",source_name=None,target_name="trim(s_address)")```
For the given example,
the source transformation is None,so the raw value in the source is considered for reconciliation.

### Do's and Don't's on Table Config:

1. Always the column reference to be source column names in all the configs,except **Transformations** and **Filters** as these

>    are expressions ,that are applied directly in the sql.

Reviewer: ```suggestion
   are sql expressions ,that are applied directly in the sql.
```

Author: done

# Guidance for Oracle as source

## Driver

> 1. **Download `ojdbc8.jar` from Oracle:**

Reviewer: we might modify this section, since we are going to execute as a job through cli, `ojdbc8.jar` will be take care automatically,

@bishwajit-db  or @vijaypavann-db  to take care while creating the cli and when source is oracle

Reviewer: Installation should take care of `ojdbc8.jar`through maven co-ordinates.

| snowflake   | array         | array_to_string(array_compact(<col_name>),',')                       | concat_ws(',', <col_name>)                      |                      |                      | in case of removing "undefined" during migration(converts sparse array to dense array)      |
| snowflake   | array         | array_to_string(array_sort(array_compact(<col_name>),true,true),',') | concat_ws(',', <col_name>)                      |                      |                      | in case of removing "undefined" during migration and want to sort the array                 |
| snowflake   | timestamp_ntz | date_part(epoch_second,<col_name>)                                   | unix_timestamp(<col_name>)                      |                      |                      | convert timestamp_ntz to epoch for getting a match b/w snowflake and databricks             | 

## Data Reconciliation Examples:

For more Reconciliation Config examples ,please refer to [sample_notebook][link].

[link]: ./docs/reconciliation-configs-examples.py

> ## Note:

Reviewer: ```suggestion
## FAQs:
```

**In Reconciliation** 
1. Databricks with UC is the only target
2. Duplicates are not covered
3. User Transformations are not applied for Schema Validation.Only select_columns,drop_columns and column_mapping is valid for schema  
validation
4. Aggregate transformations or multi-column transformations are not supported

> 5. Not all complex data types are supported

Reviewer: please add details on how various `report type` behaves under this section.

Author: Added at the top level
